### ISSUE TITLE
Avoid relying on the str() of a class object

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5349]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)| @pitrou reported|
|-|-|
|Trac ID|trac#5349|
|Type|enhancement|
|Created|2011-11-01 14:49:55Z|
|Branch|https://github.com/twisted/twisted/tree/pubkey-check-5349|

In conch.ssh, Key.type() relies on the string representation of a PyCrypto class name. This is unreliable as it can change from one Python version to another, or between implementations.
Attached patch uses the module name instead, which is reliable.

Attachments:

* [pycrypto_class_str.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f0d/f0d6caf672956573ed792304102375c6702ff87f/85884449f78ea56036e02c8cd6e0399fd0e29dad.patch) (688 bytes) - added by antoine on 2011-11-01 14:50:10Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5349 5349
type__enhancement enhancement
reporter__antoine antoine
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_pubkey_check_5349 branches/pubkey-check-5349
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1320158995000000 1320158995000000
changetime__1320183443000000 1320183443000000
version__None None
owner__exarkun exarkun
cc__z3p cc__antoine
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @exarkun|
|-|-|

This looks like an improvement, but even more of an improvement would be to use `isinstance` instead of playing games with module and class name prefixes.  The relevant objects are all instances of `Crypto.PubKey.pubkey.pubkey`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [33043]) Branching to 'pubkey-check-5349'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Alright.  My mistake for thinking there might be a better way to inspect a PyCrypto object to determine its type.  This seems to be the way to go after all.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [33044]) Apply pycrypto_class_str.patch

refs [#5349](https://github.com/twisted/twisted/issues/5349)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [33046]) Merge pubkey-check-5349

Author: antoine
Reviewer: exarkun
Fixes: [#5349](https://github.com/twisted/twisted/issues/5349)

Explicitly use the module name of the class of the key object
to determine what type of key it is, instead of relying on the
string representation of the class object.

This avoids a problem on Python 3 where the PyCrypto key class
becomes new-style and has a different string representation.
